core/lib/queued_job_processor/src/lib.rs

```diff
@@ -86,15 +86,18 @@ pub trait JobProcessor: Sync + Send {
                 );
                 let task = self.process_job(&job_id, job, started_at).await;
 
-                self.wait_for_task(job_id, started_at, task)
+                self.wait_for_task(job_id, started_at, task, &mut stop_receiver)
                     .await
                     .context("wait_for_task")?;
             } else if iterations_left.is_some() {
                 tracing::info!("No more jobs to process. Server can stop now.");
                 return Ok(());
             } else {
                 tracing::trace!("Backing off for {} ms", backoff);
-                sleep(Duration::from_millis(backoff)).await;
+                // Error here corresponds to a timeout w/o `stop_receiver` changed; we're OK with this.
+                tokio::time::timeout(Duration::from_millis(backoff), stop_receiver.changed())
+                    .await
+                    .ok();
```

Comment:
Once timeouts became errors, it turned out that job processor doesn't really respect stop signals.
While this is not a full solution, I think it's 90% of solution, which is good enough for now.